cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,548 @@
+   * Sanity Check: The total number of leader replicas of certain topics is sufficient so that these leader replicas
+   * can be moved to satisfy this goal.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    _mustHaveLeaderReplicaPerBrokerTopics = Collections.unmodifiableSet(
+        Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty()) {
+      return;
+    }
+    // Sanity checks
+    validateTopicsWithMinLeaderReplicaIsNotExcluded(clusterModel, optimizationOptions);
+    validateEnoughLeaderReplicaToDistribute(clusterModel, optimizationOptions);
+    validateBrokersAllowedReplicaMoveExist(clusterModel, optimizationOptions);
+    // Sort leader replicas for each broker based on resource utilization.
+    new SortedReplicasHelper().addPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants())
+                              .trackSortedReplicasFor(replicaSortName(this, false, false), clusterModel);
+  }
+
+  private int minTopicLeadersPerBroker() {
+    return _balancingConstraint.minTopicLeadersPerBroker();
+  }
+
+  private void validateTopicsWithMinLeaderReplicaIsNotExcluded(ClusterModel clusterModel,
+                                                               OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty() || optimizationOptions.excludedTopics().isEmpty()) {
+      return;
+    }
+    Set<String> shouldNotBeExcludedTopics = new HashSet<>();
+    _mustHaveLeaderReplicaPerBrokerTopics.forEach(topicName -> {
+      if (optimizationOptions.excludedTopics().contains(topicName)) {
+        shouldNotBeExcludedTopics.add(topicName);
+      }
+    });
+    if (!shouldNotBeExcludedTopics.isEmpty()) {
+      throw new OptimizationFailureException(String.format("Topics that must have a minimum number of leader replicas "
+          + "per alive broker that is not excluded for leadership moves should not be excluded. This error implies"
+          + "a config error. Topics should not be excluded=[%s]", String.join(", ", shouldNotBeExcludedTopics)));
+    }
+  }
+
+  private void validateEnoughLeaderReplicaToDistribute(ClusterModel clusterModel,
+                                                       OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty()) {
+      return;
+    }
+    Map<String, Set<Replica>> replicasByTopicNames = clusterModel.replicasOf(_mustHaveLeaderReplicaPerBrokerTopics);
```

Comment:
Looks like the only use for this variable is to count the number of leaders for each relevant topic. This makes the creation of `clusterModel#replicasOf` and the following use of `this#countLeaderReplicas` inefficient / redundant.

Instead, does it make sense to
1. Eliminate `clusterModel#replicasOf`,
2. Create `clusterModel#numLeadersPerTopic(Set<String> topics)`, which returns a `Map<String, Integer>`, showing number of leaders per given topic?

Reply:
Good idea